### PR TITLE
Add gx-map-line control

### PR DESCRIPTION
**Pull request by fedes1987**
gx-map-line control was added. It allows create polylines in the gx-map.

### CODE REVIEW

File: src/components/common/coordsValidate.ts [hunk 1]
```diff
@@ -1,5 +1,46 @@
...
+  if (result !== null) {
+    return result.slice(1);
+  } else {
+    //it's geopoint/geostring type
+    let geoObject;
+    try {
+      geoObject = JSON.parse(coord);
+    } catch (e) {
+      throw new Error("Invalid GeoJSON object.");
+    }
```
Comment (ncardeli) [inline src/components/common/coordsValidate.ts:13]: Maybe we should allow the original error to throw here, to let the programmer understand the cause of the error, that in this case would be a parsing error.

File: src/components/common/coordsValidate.ts [hunk 2]
```diff
@@ -1,5 +1,46 @@
...
+      } else {
+        const isSDTArray =
+          geoObject[0].hasOwnProperty("Geography") && Array.isArray(geoObject);
+        if (isSDTArray) {
+          geoObject["IsLayersArray"] = true;
+          return geoObject;
+        }
+      }
+    }
+    return geoObject;
```
Comment (ncardeli) [inline src/components/common/coordsValidate.ts:44]: Maybe we could extract this logic to a method focused on parsing a GeoJson.
What do you think? That way, the comment would not be needed, and this block of code would be self-described by the method name.

File: src/components/common/coordsValidate.ts [hunk 3]
```diff
@@ -1,5 +1,46 @@
...
+      }
+    } else {
+      const hasGeography = geoObject.hasOwnProperty("Geography");
+      if (hasGeography) {
+        return geoObject;
+      } else {
+        const isSDTArray =
+          geoObject[0].hasOwnProperty("Geography") && Array.isArray(geoObject);
+        if (isSDTArray) {
+          geoObject["IsLayersArray"] = true;
```
Comment (ncardeli) [inline src/components/common/coordsValidate.ts:39]:
```suggestion
          geoObject.IsLayersArray = true;
```

File: src/components/common/coordsValidate.ts [hunk 4]
```diff
@@ -1,5 +1,46 @@
...
+        default:
+          throw new Error('Invalid GeoJSON object.');
+          */
+      }
+    } else {
+      const hasGeography = geoObject.hasOwnProperty("Geography");
+      if (hasGeography) {
+        return geoObject;
+      } else {
+        const isSDTArray =
```
Comment (ncardeli) [inline src/components/common/coordsValidate.ts:36]: How about `isArray`?

In the context of this library, SDT has no meaning.

File: src/components/map/map.tsx [hunk 5]
```diff
@@ -169,6 +170,23 @@ export class Map implements GxComponent {
...
+      this.element.addEventListener("gxMapDidLoad", () => {
+        lineV.addTo(this.map);
+      });
+    }
+
+    lineElement.addEventListener("gxMapLineDeleted", () => {
+      this.onMapLineDeleted(lineV);
+    });
+  }
+
```
Comment (ncardeli) [inline src/components/map/map.tsx:189]: Looks good!

File: src/components/map/map.tsx [hunk 6]
```diff
@@ -226,6 +244,24 @@ export class Map implements GxComponent {
     }
   }
 
+  private onMapLineDeleted(line: polyline) {
+    //const i = 0;
```
Comment (ncardeli) [inline src/components/map/map.tsx:248]: As a convention, we avoid as much as possible to commit commented code.

File: src/components/map-line/map-line.tsx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  h,
+  Prop
+} from "@stencil/core";
+import { Component as GxComponent } from "../common/interfaces";
+import { polyline } from "leaflet/dist/leaflet-src.esm";
+import { parseCoords } from "../common/coordsValidate";
+
+@Component({
+  shadow: false,
+  //styleUrl: "map-line.scss",
```
Comment (ncardeli) [inline src/components/map-line/map-line.tsx:15]: Let's remove this commented line as it doesn't add anything useful 😀

File: src/components/map-line/map-line.tsx [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  /**
+   * Emmits when the element is deleted from a `<gx-map>`.
+   *
+   */
+  @Event() gxMapLineDeleted: EventEmitter;
+
+  private setupLine(coords) {
+    this.lineInstance = polyline(coords, {
+      color: "red",
+      weight: 3
```
Comment (ncardeli) [inline src/components/map-line/map-line.tsx:43]: I feel like these two should be properties of the component, to allow customization, but we can improve this in a future PR.

File: src/components/map-line/map-line.tsx [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      );
+      this.setupLine([0, 0]);
+    }
+    /*
+        this.setPopup();
+        if (this.tooltipCaption) {
+          this.markerInstance.bindTooltip(this.tooltipCaption, {
+            direction: "top"
+          });
+        }*/
```
Comment (ncardeli) [inline src/components/map-line/map-line.tsx:64]: Let's erase this commented code.

File: src/components/map-line/map-line.tsx [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    this.gxMapLineDidLoad.emit(this.lineInstance);
+  }
+
+  componentDidUnload() {
+    this.gxMapLineDeleted.emit(this.lineInstance);
+  }
+
+  render() {
+    return (
+      <div class="popup-data-container">
```
Comment (ncardeli) [inline src/components/map-line/map-line.tsx:74]: What is this class used for? 🤔

File: src/components/map-line/readme.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
+# gx-map-line
+
```
Comment (ncardeli) [inline src/components/map-line/readme.md:2]: A couple of lines explaining what the component does would be great!
Bonus points if you paste a usage example!

**Review by ncardeli — CHANGES_REQUESTED** [review]
Excellent work! This looks like a really nice addition to the `gx-map` component.
Left some comments for you.

File: src/components/common/coordsValidate.ts (hunk 1 above)
Comment (fedes1987) [reply src/components/common/coordsValidate.ts]: Done. Addresed in 6aa1ac5

File: src/components/common/coordsValidate.ts (hunk 3 above)
Comment (fedes1987) [reply src/components/common/coordsValidate.ts]: Done. Addresed in 6aa1ac5

File: src/components/common/coordsValidate.ts (hunk 4 above)
Comment (fedes1987) [reply src/components/common/coordsValidate.ts]: Done. Addresed in 6aa1ac5

File: src/components/map-line/map-line.tsx (hunk 7 above)
Comment (fedes1987) [reply src/components/map-line/map-line.tsx]: Done. Addresed in 6aa1ac5

File: src/components/map-line/map-line.tsx (hunk 9 above)
Comment (fedes1987) [reply src/components/map-line/map-line.tsx]: Done. Addresed in 6aa1ac5

File: src/components/map-line/map-line.tsx (hunk 10 above)
Comment (fedes1987) [reply src/components/map-line/map-line.tsx]: Honestly, no idea. It was a copy/pasted code. I removed it.
Addresed in 6aa1ac5

File: src/components/map-line/readme.md (hunk 11 above)
Comment (fedes1987) [reply src/components/map-line/readme.md:2]: Done. Addresed in 6aa1ac5

File: src/components/map-line/map-line.tsx (hunk 8 above)
Comment (fedes1987) [reply src/components/map-line/map-line.tsx:42]: I agree.

File: src/components/map/map.tsx (hunk 6 above)
Comment (fedes1987) [reply src/components/map/map.tsx]: Finally I decided to implement the linesList and the full method OnMapLineDeleted.
Addressed in b9c59f7

File: src/components/common/coordsValidate.ts (hunk 2 above)
Comment (ncardeli) [reply src/components/common/coordsValidate.ts]: Did would be a nice improvement and would make the code more readable. What do you think @fedes1987 ?

File: src/components/map-line/map-line.tsx [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    });
+  }
+
+  componentDidLoad() {
+    const coords = parseCoords(this.coords);
+    if (coords !== null) {
+      this.setupLine(coords);
+    } else {
+      console.warn(
+        "GX warning: Can not read 'coords' attribute, default coords set (gx-map-line)",
```
Comment (ncardeli) [inline src/components/map-line/map-line.tsx:52]:
```suggestion
        "GX warning: Cannot read 'coords' attribute, default coords set (gx-map-line)",
```

**Review by ncardeli — COMMENTED** [review]
Almost there! I'd like your input on the comment I left about extracting some of the GeoJson validation to a new function.

File: src/components/common/coordsValidate.ts (hunk 2 above)
Comment (fedes1987) [reply src/components/common/coordsValidate.ts]: I agree. 
Addressed in 33b256a

**Review by ncardeli — APPROVED** [review]
💯

**Pull request merged**